src/mam4xx/aero_model.hpp

```diff
@@ -0,0 +1,626 @@
+// Battelle Memorial Institute and
+// National Technology & Engineering Solutions of Sandia, LLC (NTESS)
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef MAM4XX_AERO_MODEL_HPP
+#define MAM4XX_AERO_MODEL_HPP
+
+#include <ekat/util/ekat_math_utils.hpp>
+
+#include <haero/atmosphere.hpp>
+#include <haero/math.hpp>
+
+#include <mam4xx/aero_config.hpp>
+#include <mam4xx/conversions.hpp>
+#include <mam4xx/mam4_types.hpp>
+#include <mam4xx/utils.hpp>
+
+namespace mam4 {
+
+namespace aero_model {
+
+// BAD CONSTANT
+const Real dlndg_nimptblgrow = haero::log(1.25);
+constexpr int nimptblgrow_mind = -7, nimptblgrow_maxd = 12;
+constexpr int nimptblgrow_total = -nimptblgrow_mind + nimptblgrow_maxd + 1;
+const int nrainsvmax = 50; // maximum bin number for rain
+const int naerosvmax = 51; //  maximum bin number for aerosol
+const int maxd_aspectype = 14;
+
+KOKKOS_INLINE_FUNCTION
+void modal_aero_bcscavcoef_get(
+    const int imode, const bool isprx_kk, const Real dgn_awet_imode_kk, //& ! in
+    const Real dgnum_amode_imode,
+    const Real scavimptblvol[nimptblgrow_total][AeroConfig::num_modes()],
+    const Real scavimptblnum[nimptblgrow_total][AeroConfig::num_modes()],
+    Real &scavcoefnum_kk, Real &scavcoefvol_kk) {
+
+  // !-----------------------------------------------------------------------
+  // ! compute impaction scavenging removal amount for aerosol volume and number
+  // !-----------------------------------------------------------------------
+
+  // @param [in]  imode mode index
+  // @param [in]  isprx_kk if there is precip
+  // @param [in]  dgn_awet_imode_kk ! wet aerosol diameter of mode imode at
+  // elevation kk [m]
+  // @param [out] scavcoefnum scavenging removal for aerosol number [1/h]
+  // @param [out] scavcoefvol scavenging removal for aerosol volume [1/h]
+
+  // NOTE: original FORTRAN function has two internal loops: kk and icol.
+  // We removed these loops. Hence, the inputs/outputs of
+  // modal_aero_bcscavcoef_get are reals at kk , icol locations
+
+  const Real zero = 0;
+  const Real one = 1;
+  if (isprx_kk) {
+    // ! interpolate table values using log of
+    // (actual-wet-size)/(base-dry-size) ratio of wet and dry aerosol diameter
+    // [fraction]
+    const Real wetdiaratio = dgn_awet_imode_kk / dgnum_amode_imode;
+    // BAD CONSTANT
+    Real scavimpvol, scavimpnum = zero;
+    if (wetdiaratio >= 0.99 && wetdiaratio <= 1.01) {
+      scavimpvol = scavimptblvol[-nimptblgrow_mind][imode];
+      scavimpnum = scavimptblnum[-nimptblgrow_mind][imode];
+    } else {
+      Real xgrow = haero::log(wetdiaratio) / dlndg_nimptblgrow;
+      int jgrow = int(xgrow);
+      if (xgrow < zero) {
+        jgrow = jgrow - 1;
+      }
+
+      if (jgrow < nimptblgrow_mind) {
+        jgrow = nimptblgrow_mind;
+        xgrow = jgrow;
+      } else {
+        jgrow = haero::min(jgrow, nimptblgrow_maxd - 1);
+      }
+      const Real dumfhi = xgrow - jgrow;
+      const Real dumflo = one - dumfhi;
```

Comment:
If either or both of you understand what's happening here, it would be nice to add some comments.

Reply:
maybe @pressel, @singhbalwinder, and @tangshuaiqi could help us with this.

Reply:
The whole purpose of the code above is to find `jgrow`, which is an index for arrays `scavimptblvol` and `scavimptblnum`. We compute `xgrow` (a real number) and take `int` of it to get our index `jgrow`. Since `xgrow` can be negative, we change jgrow appropriately (the `int` function rounds towards zero in both Fortran and C).  

After that, we are ensuring that jgrow is within the bounds of `nimptblgrow_maxd` and `nimptblgrow_mind`. We compute `dumfhi` and `dumflo` to some sort of interpolation.

We can add comments similar to the following in the code:

```
      Real xgrow = haero::log(wetdiaratio) / dlndg_nimptblgrow;
      int jgrow = int(xgrow); // get index jgrow
      if (xgrow < zero) { // adjust jgrow appropriately if xgrow is negative
        jgrow = jgrow - 1;
      }
      // bound jgrow within max and min values
      if (jgrow < nimptblgrow_mind) {
        jgrow = nimptblgrow_mind;
        xgrow = jgrow;
      } else {
        jgrow = haero::min(jgrow, nimptblgrow_maxd - 1);
      }
     // compute factors for interpolating impaction scavenging removal amounts
      const Real dumfhi = xgrow - jgrow;
      const Real dumflo = one - dumfhi;
```

Reply:
@odiazib or @mjs271 , can you add @singhbalwinder 's suggested comments (or something like them) to this to the code above?

Reply:
Done.